### PR TITLE
Large-Scale refactoring

### PR DESCRIPTION
**Pull request by Vogel612**
- Changes to Package-Structure
- Additional Documentation for Bot Core Classes
- minor POISON_PILL changes
- changes in message processing by streams
- Got rid of excess ScheduledExecutorService.
- Minor bigfixes

### CODE REVIEW

File: src/main/java/com/gmail/inverseconduit/bot/AbstractBot.java [hunk 1]
```diff
@@ -4,24 +4,59 @@
...
+ * enqueue messages to {@link #messageQueue}.
+ * </br></br>If the {@link ChatWorker#POISON_PILL POISON_PILL} is enqueued,
+ * {@link #shutdown()} will be called to notify implementing classes of the
+ * shutdown request.
+ * Also the {@link #processingThread}'s {@link ExecutorService#shutdown()
+ * shutdown}-method will be called. It is accordingly not usable after this.
+ * 
+ * @author Vogel612<<a href="mailto:vogel612@gmx.de"
+ *         >vogel612@gmx.de</a>>
+ */
```
Comment (fge) [inline src/main/java/com/gmail/inverseconduit/bot/AbstractBot.java]: General note about javadoc formatting:

* the first line right after the `**` will appear in the summary;
* the rest will appear as the text when a user clicks on the method link.

As such, the format I usually use is this:

```java
/**
 * Some short summary of the field/method/etc
 *
 * <p>Here, any HTML tags</p>
 *
 * <ul>
 *     <li>can be used</li>
 * </ul>
 */
```

Note that `{@code ...}`, `{@link ...}` and friends also work in the summary line.

File: src/main/java/com/gmail/inverseconduit/bot/AbstractBot.java [hunk 2]
```diff
@@ -4,24 +4,59 @@
...
 public abstract class AbstractBot implements ChatWorker {
 
-    protected final ScheduledExecutorService   executor         = Executors.newSingleThreadScheduledExecutor();
+    private static ThreadLocal<Boolean>        shutdown                = ThreadLocal.withInitial(() -> false);
```
Comment (fge) [inline src/main/java/com/gmail/inverseconduit/bot/AbstractBot.java]: Make it final

File: src/main/java/com/gmail/inverseconduit/bot/AbstractBot.java [hunk 3]
```diff
@@ -4,24 +4,59 @@
...
+                                                                                           "Didn't get a message over the course of 10 minutes... something might be wrong", e);
+                                                                               }
+                                                                               if (headMessage != null) {
+                                                                                   Logger.getLogger(this.getClass().getName()).finest("processing message from queue");
+                                                                                   return headMessage;
+                                                                               }
+                                                                           }
+                                                                           // at this point, users of the supplier will have been shutdown, so it's moot
+                                                                               return null;
+                                                                           };
```
Comment (fge) [inline src/main/java/com/gmail/inverseconduit/bot/AbstractBot.java]: Make this lambda into a method and fo `= this::theSupplierMethod` instead?

File: src/main/java/com/gmail/inverseconduit/bot/DefaultBot.java [hunk 4]
```diff
@@ -53,14 +53,12 @@ public static DefaultBot create(ChatInterface chatInterface) {
...
+        // asynchronously enqueue the processing by blocking supplier
+        processingThread.submit(this::processMessageQueue);
     }
 
     private void processMessageQueue() {
-        while (messageQueue.peek() != null) {
-            LOGGER.finest("processing message from queue");
-            processingThread.submit(() -> processMessage(messageQueue.poll()));
-        }
+        Stream.generate(blockingMessageSupplier).forEach(headMessage -> processingThread.submit(() -> processMessage(headMessage)));
```
Comment (fge) [inline src/main/java/com/gmail/inverseconduit/bot/DefaultBot.java]: A classic loop here is imho more readable!

File: src/main/java/com/gmail/inverseconduit/bot/InteractionBot.java [hunk 5]
```diff
@@ -26,13 +26,12 @@ public InteractionBot(ChatInterface chatInterface) {
 
     @Override
     public void start() {
-        executor.scheduleAtFixedRate(this::processInteractions, 200, 700, TimeUnit.MILLISECONDS);
+        // asynchronously enqueue the processing by blocking supplier
+        processingThread.submit(this::processInteractions);
```
Comment (fge) [inline src/main/java/com/gmail/inverseconduit/bot/InteractionBot.java]: Same here; a `ScheduledExecutorService` conveys the intent better. Abstraction has limits ;)

File: src/main/java/com/gmail/inverseconduit/bot/DefaultBot.java (hunk 4 above)
Comment (Vogel612) [inline src/main/java/com/gmail/inverseconduit/bot/DefaultBot.java]: Reverting to a classic for loop means losing the benefit of the blockingMessageSupplier and adding yet another Thread Pool that needs to be taken care of.

While I agree that a classic loop may be more readable / conventional I personally think that the blocking wait is cleaner

File: src/main/java/com/gmail/inverseconduit/bot/InteractionBot.java (hunk 5 above)
Comment (Vogel612) [inline src/main/java/com/gmail/inverseconduit/bot/InteractionBot.java]: Maybe an appropriate rename can help here. I considered taking things a step further and using the same pattern as for shutdown by providing a default implementation for `start()` and delegate the processing to an abstract method.

File: src/main/java/com/gmail/inverseconduit/bot/AbstractBot.java (hunk 1 above)
Comment (Vogel612) [inline src/main/java/com/gmail/inverseconduit/bot/AbstractBot.java]: That Javadoc format sounds interesting. Currently I just ran the shared formatting preferences, maybe they can be adapted to include this javadoc format. Thoughts @Unihedro ?

File: src/main/java/com/gmail/inverseconduit/bot/InteractionBot.java (hunk 5 above)
Comment (fge) [inline src/main/java/com/gmail/inverseconduit/bot/InteractionBot.java]: However, this abstraction fails to convey the fact that this "blocking" is in fact periodic; that's my point.

Comment (Vogel612) [inline src/main/java/com/gmail/inverseconduit/bot/InteractionBot.java]: So you'd rather have it wait indefinitely? I'll go look through the Queues if I can find something that blocks non-periodically and use that instead, then

Comment (fge) [inline src/main/java/com/gmail/inverseconduit/bot/InteractionBot.java]: Wait wait wait; I don't know the intent behind the code there; but if indeed a timer is used for polling then a `BlockingQueue<?>` would indeed be more appropriate here, and again, that's imho.

Comment (Vogel612) [inline src/main/java/com/gmail/inverseconduit/bot/InteractionBot.java]: There is a blocking queue. I changed the blocking supplier from `poll()` to `take()`, which block until an element becomes available. These elements are then streamed infinitely and enqueued into a separate thread to be processed.

File: src/main/java/com/gmail/inverseconduit/commands/sets/CoreBotCommands.java [hunk 6]
```diff
@@ -15,15 +15,16 @@
 
 import com.gmail.inverseconduit.AppContext;
 import com.gmail.inverseconduit.BotConfig;
-import com.gmail.inverseconduit.bot.Program;
 import com.gmail.inverseconduit.chat.ChatInterface;
+import com.gmail.inverseconduit.chat.ChatWorker;
 import com.gmail.inverseconduit.chat.Subscribable;
 import com.gmail.inverseconduit.commands.CommandHandle;
+import com.gmail.inverseconduit.commands.providers.canned_comments.CannedComments;
```
Comment (Vogel612) [inline src/main/java/com/gmail/inverseconduit/commands/sets/CoreBotCommands.java]: this line prevents compilation of the pull-request currently. It slipped in when I didn't have this and https://github.com/Vogel612/JavaBot/tree/canned-comments separate yet

**Pull request merged**